File: 13/d13-1.py
import json

def compare(lhs, rhs):
    """Compares two lists"""
    if isinstance(lhs, int) and isinstance(rhs, int):
        diff = lhs - rhs
        return diff // abs(diff) if diff else 0
    if isinstance(lhs, list) and isinstance(rhs, int):
        res = compare(lhs, [rhs])
        return res
    if isinstance(lhs, int) and isinstance(rhs, list):
        res = compare([lhs], rhs)
        return res
    for ll, rr in zip(lhs, rhs):
        comp = compare(ll, rr)
        if comp:
            return comp
    diff = len(lhs) - len(rhs)
    return diff // abs(diff) if diff else 0

ind = 0
indsum = 0
fp = open("input.dat", "r")
while True:
    line1 = fp.readline().strip()
    line2 = fp.readline().strip()
    fp.readline()
    if not line1 or not line2:
        break
    ind += 1
    lhs = json.loads(line1)
    rhs = json.loads(line2)
    if compare(lhs, rhs) != 1:
        indsum += ind
fp.close()

print(f"Index sum of packages in right order: {indsum}")

File: 13/d13-1.f90
program d13_1
  use d13_helper, only : build_tree, compare_trees, tree_node
  implicit none

  integer, parameter :: max_line_len = 1024

  integer :: fd, iostat, icycle, indsum
  character(max_line_len) :: line
  type(tree_node), allocatable :: tree1, tree2

  indsum = 0
  icycle = 0
  open(newunit=fd, file="input.dat")
  do
    icycle = icycle + 1
    read(fd, "(a)", iostat=iostat) line
    if (iostat /= 0) exit
    call build_tree(trim(line), tree1)
    read(fd, "(a)", iostat=iostat) line
    if (iostat /= 0) exit
    call build_tree(trim(line), tree2)
    if (compare_trees(tree1, tree2) <= 0) indsum = indsum + icycle
    read(fd, *, iostat=iostat)
    if (iostat /= 0) exit
  end do
  close(fd)

  print "(a, i0)", "Index sum of packages in right order: ", indsum

end program d13_1
